src/match.js

```diff
@@ -55,20 +65,54 @@ function doesUnitMatch(expr, str) {
   const [head, operator, rest] = splitExpr(expr);
   if (isLiteral(head)) {
     return expr[0] === str[0];
+  } else if (isDot(head)) {
+    return true;
   } else if (isSet(head)) {
     const set_terms = splitSet(head);
     return set_terms.includes(str[0]);
   }
   return false;
 }
 
+function matchMultiple(expr, str, match_length, min_length = 0, max_length = -1) {
+  [head, operator, rest] = splitExpr(expr);
+  let submatch_length = -1;
+
+  while (max_length === -1 || (submatch_length < max_length)) {
+    const [subexpr_matched] = matchExpr(
+      duplicate(head, submatch_length + 1), 
+      str, 
+      match_length
+    );
+    if (!subexpr_matched) break;
+    submatch_length++;
+  }
+
+  while (submatch_length >= min_length) {
+    const [matched, new_match_length] = matchExpr(
+        duplicate(head, submatch_length) + rest, 
+        str, 
+        match_length
+    );
+    if (matched) return [matched, new_match_length];
+    submatch_length--;
+  }
+
+  return [false, 0];
+}
```

Comment:
Since this function is quite long, it would be great to add some comments to explain what each part does. Other code is good!

Reply:
In the future, when I end realization of business logic, I will decompose and use prettier to get pure code.